driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,247 @@
+use super::stablex_driver::StableXDriver;
+use crate::metrics::StableXMetrics;
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::{anyhow, Context, Result};
+use log::error;
+use log::info;
+use std::collections::HashSet;
+use std::thread;
+use std::time::{Duration, Instant, SystemTime, SystemTimeError};
+
+const BATCH_DURATION: Duration = Duration::from_secs(300);
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct BatchId(pub u64);
+
+impl BatchId {
+    pub fn current(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        let time_since_epoch = now.duration_since(SystemTime::UNIX_EPOCH)?;
+        Ok(Self(time_since_epoch.as_secs() / BATCH_DURATION.as_secs()))
+    }
+
+    pub fn currently_being_solved(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        Self::current(now).map(|batch_id| batch_id.prev())
+    }
+
+    pub fn start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    pub fn solve_start_time(self) -> SystemTime {
+        self.start_time() + BATCH_DURATION
+    }
+
+    pub fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    pub fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    orderbook_reader: &'a dyn StableXOrderBookReading,
+    metrics: &'a StableXMetrics,
+    batch_wait_time: Duration,
+    max_batch_elapsed_time: Duration,
+    past_auctions: HashSet<BatchId>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum Action {
+    AlreadyHandled,
+    NotEnoughTimeLeft,
+    Solve,
+}
+
+impl<'a> Scheduler<'a> {
+    pub fn new(
+        driver: &'a mut dyn StableXDriver,
+        orderbook_reader: &'a dyn StableXOrderBookReading,
+        metrics: &'a StableXMetrics,
+        batch_wait_time: Duration,
+        max_batch_elapsed_time: Duration,
+    ) -> Self {
+        Self {
+            driver,
+            orderbook_reader,
+            metrics,
+            batch_wait_time,
+            max_batch_elapsed_time,
+            past_auctions: HashSet::new(),
+        }
+    }
+
+    pub fn run_forever(&mut self) {
+        loop {
+            let now = SystemTime::now();
```

Comment:
Do we have to be careful with timezones here? Can we specify we want now in UTC somehow?

Reply:
From what I understand from the docs, `SystemTime` is completely opaque and cannot be directly inspected, but rather compared to another time. Since `UNIX_EPOCH` is timezone agnostic, I don't think its an issue here.

Reply:
I came to the same conclusion. Time zones don't affect this.